### PR TITLE
fix(core-utils): more resiliently expand flex mode

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
Part of https://github.com/opentripplanner/otp-react-redux/pull/530

Small adjustment to how we replace `FLEX` with the 3 OTP flex modes. The new approach is more computationally expensive, but more resilient in more situations.

### CODE REVIEW

**Review by daniel-heppner-ibigroup — APPROVED** [review]

File: packages/core-utils/src/query.js [hunk 1]
```diff
@@ -466,8 +466,19 @@ export function getRoutingParams(config, currentQuery, ignoreRealtimeUpdates) {
...
   if (params.mode) {
     // Ensure query is in reduced format to avoid replacing twice
-    const mode = reduceOtpFlexModes(params.mode.split(",")).join(",");
-    params.mode = mode.replace("FLEX", "FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT");
+    const mode = reduceOtpFlexModes(params.mode.split(","));
+    params.mode = mode
+      .map(m => {
+        // If both the expanded and shrunk modes are included, remove the exapnded one
+        if (m === "FLEX_EGRESS" || m === "FLEX_ACCESS" || m === "FLEX_DIRECT") {
+          if (mode.includes("FLEX")) return "";
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query.js:474]: Is this going to result in an OTP mode list where there is a comma followed by nothing or another comma?

File: packages/core-utils/src/query.js [hunk 2]
```diff
@@ -466,8 +466,19 @@ export function getRoutingParams(config, currentQuery, ignoreRealtimeUpdates) {
...
+        }
+        if (m === "FLEX") {
+          return "FLEX_EGRESS,FLEX_ACCESS,FLEX_DIRECT";
+        }
+        return m;
+      })
+      .join(",");
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query.js:481]: You should extract this changed block into its own method and add a test so we can see the expected inputs/outputs.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
I think a method should be created to hide that logic, and tests should be added to document the expected behavior.

File: packages/core-utils/src/query.js (hunk 1 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/query.js:474]: This is checked by the new unit tests and seems to work correctly

File: packages/core-utils/src/query.js (hunk 2 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/query.js:481]: Good call! Lots of unit tests added in 70d83757ca12f50aea4417d847732854c6c660b8

File: packages/core-utils/src/__tests__/query-params.js [hunk 3]
```diff
@@ -38,3 +39,51 @@ describe("query-params", () => {
...
 });
+
+describe("flex-reducer", () => {
+  it("should not touch a query that doesn't include flex modes", () => {
+    expect(reduceOtpFlexModes(["WALK", "TRANSIT", "BIKE"])).toMatchSnapshot();
+  });
+  it("should modify a query that includes some flex modes", () => {
+    expect(
+      reduceOtpFlexModes(["WALK", "TRANSIT", "BIKE", "FLEX_DIRECT"])
+    ).toMatchSnapshot();
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/__tests__/query-params.js:50]: You probably don't need snapshots for simple object comparison. Use `expect(...).toBe(<expected_value>)`.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Thanks for adding the tests, see https://github.com/opentripplanner/otp-ui/pull/332/files/095d0131000373cfc065d33cfcefecd5260cb820..d942adf911c22009318496f5c71facfcd836ca84#r792690034 to simplify.

File: packages/core-utils/src/__tests__/query-params.js (hunk 3 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/__tests__/query-params.js]: Snapshots moved in 7be6cdf4055042904e0d895d22825d5df980920f

**Comment by miles-grant-ibigroup** [comment]
Thanks Binh, unit tests refactored

File: packages/core-utils/src/__tests__/query.js [hunk 4]
```diff
@@ -102,6 +102,58 @@ describe("query", () => {
...
       ).toMatchSnapshot();
     });
+
+    it("should create routing params for standard flex query", () => {
+      expect(
+        getRoutingParams(fakeConfig, {
+          ...makeBaseTestQuery(),
+          mode: "WALK,TRANSIT,FLEX"
+        })
+      ).toMatchSnapshot();
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/__tests__/query.js:112]: Those are probable fine as snapshots.

**Review by binh-dam-ibigroup — APPROVED** [review]
Looks good!

**Pull request merged**